Restructure MFA backend to support multiple devices

awly opened the pull request:
Each user can now have multiple devices. This commit only changes the
backend structure to support it, the client and API haven't been updated
yet.

Also added a migration for existing MFA data on auth server startup.

### notes on protobufs and json

`Makefile` has changes in proto generation - using `--gogofast_out` instead of `--gofast_out`. This results in the generated stubs importing `github.com/gogo/protobuf/proto` instead of `github.com/golang/protobuf/proto`, which in turn makes them compatible with https://godoc.org/github.com/gogo/protobuf/jsonpb.

`jsonpb` has to be used instead of `encoding/json` because of the `oneof` field in the proto. The `encoding/json` package can't unmarshal into the proto-generated structs with `oneof`, because those get generated as interface fields.

One downside is that `jsonpb` doesn't respect the `json` struct tags, so all the names in serialized JSON match the Go struct field names.
I tried using https://godoc.org/google.golang.org/protobuf/encoding/protojson instead, but that doesn't work on gogoproto-genereated stubs.

### other notes

JSON schema for `LocalAuth` got pretty deep and could be broken up with more `Sprintf` nesting. But I plan on just ripping it all out anyway https://github.com/gravitational/teleport/issues/5402, so keeping as is.

There was also a stray uuid dependency in our `vendor/`, which got cleaned up via `make update-vendor`. It's unrelated to this PR.

a-palchikov commented:
Regarding jsonpb ignoring json tags - I've looked at the package, specifically into the tests.
Curiously, the generated protobuf file includes a `protobuf` property called `json` which specifies the name of the resulting JSON attribute name although the test protobuf is using version 2 and has not applied any gogo-specific options: https://github.com/gogo/protobuf/blob/b03c65ea87cdc3521ede29f62fe3ce239267c1bc/jsonpb/jsonpb_test.go#L80 and https://github.com/gogo/protobuf/blob/b03c65ea87cdc3521ede29f62fe3ce239267c1bc/jsonpb/jsonpb_test_proto/test_objects.pb.go#L69 (`oBool`)

a-palchikov commented on api/types/authentication.go:
```diff
@@ -295,121 +296,263 @@ type U2F struct {
+// Equals checks equality (nil safe).
+func (l *LocalAuthSecrets) Equals(other *LocalAuthSecrets) bool {
+	if (l == nil) || (other == nil) {
+		return (l == nil) && (other == nil)
```
`return l == other`

a-palchikov commented on api/types/authentication.go:
```diff
@@ -295,121 +296,263 @@ type U2F struct {
+		return trace.Wrap(err)
+	}
+	if d.Kind == "" {
+		return trace.BadParameter("MFADevice missing ID field")
```
```suggestion
		return trace.BadParameter("MFADevice missing Kind field")
```

a-palchikov commented on api/types/authentication.go:
```diff
@@ -295,121 +296,263 @@ type U2F struct {
+	return jsonpb.Unmarshal(bytes.NewReader(buf), d)
+}
+
+// NewU2FDevice creates a TOTP MFADevice from the given key.
```
```suggestion
// NewTOTPDevice creates a TOTP MFADevice from the given key.
```

a-palchikov commented on api/types/authentication.go:
```diff
@@ -295,121 +296,263 @@ type U2F struct {
-			return trace.Wrap(err)
-		}
+	if _, err := d.GetPubKeyDecoded(); err != nil {
+		return trace.BadParameter("U2fDevice PubKey is invalid: %v", err)
```
```suggestion
		return trace.BadParameter("U2FDevice PubKey is invalid: %v", err)
```
for consistency with other messages.

a-palchikov commented on api/types/authentication.go:
```diff
@@ -460,7 +603,48 @@ const LocalAuthSecretsSchema = `{
+					"subKind": {"type": "string"},
+					"version": {"type": "string"},
+					"metadata": {
+						"type": "object",
```
nit: this should just reference the metadata spec if the JSON schema implementation is capable of references to avoid duplication.

a-palchikov commented on lib/auth/auth.go:
```diff
@@ -854,15 +869,29 @@ func (a *Server) CheckU2FSignResponse(user string, response *u2f.SignResponse) e
 		return trace.Wrap(err)
 	}
+	// TODO(awly): support multiple U2F device challenges.
+	// TODO(awly): support multiple U2F device challenges.
```
```suggestion
```

a-palchikov commented on lib/auth/init_test.go:
```diff
@@ -387,3 +399,132 @@ func (s *AuthInitSuite) TestCASigningAlg(c *C) {
+	}
+	cmpOpts := []cmp.Option{
+		cmpopts.IgnoreFields(types.UserSpecV2{}, "CreatedBy"),
+		cmpopts.IgnoreFields(types.MFADevice{}, "Id", "AddedAt", "LastUsed"),
```
I'd not ignore the time fields and instead used a fake time anchor

a-palchikov commented on lib/auth/password_test.go:
```diff
@@ -207,8 +209,11 @@ func (s *PasswordSuite) TestChangePasswordWithOTP(c *C) {
 	otpSecret := base32.StdEncoding.EncodeToString([]byte("def456"))
-	err = s.a.UpsertTOTP(req.User, otpSecret)
-	c.Assert(err, IsNil)
+	dev, err := types.NewTOTPDevice("otp", otpSecret, s.a.clock.Now())
```
The clock on the auth server is reset below - does the clock value need to be lifted?

awly replied:
done

Commit:
u2f: restructure MFA backend to support multiple devices

Each user can now have multiple devices. This commit only changes the
backend structure to support it, the client and API haven't been updated
yet.

Also added a migration for existing MFA data on auth server startup.

r0mant commented on api/types/authentication.go:
```diff
@@ -295,121 +296,263 @@ type U2F struct {
-		return trace.BadParameter("missing u2f key handle")
+// Equals checks equality (nil safe).
+func (l *LocalAuthSecrets) Equals(other *LocalAuthSecrets) bool {
+	if (l == nil) || (other == nil) {
```
Nit: Is this the same as 

```go
if l == nil && other == nil {
  return true
}
```

? :) Would be a bit easier to read IMO.

r0mant replied:
Ah, right. Bool math failure on my part 🤦

r0mant commented on api/types/types.proto:
```diff
@@ -1154,15 +1150,48 @@ message LocalAuthSecrets {
-    // U2FRegistration holds Universal Second Factor registration info.
     U2FRegistrationData U2FRegistration = 3 [ (gogoproto.jsontag) = "u2f_registration,omitempty" ];
-
-    // U2FCounter holds the highest seen Universal Second Factor registration
```
Did you remove godocs on purpose?

awly replied:
Yes, these 3 fields are deprecated and only exist to allow the auth server to load old data from the backend and migrate it to `MFA`

r0mant commented on lib/auth/auth.go:
```diff
@@ -843,11 +843,28 @@ func (a *Server) CheckU2FSignResponse(user string, response *u2f.AuthenticateCha
+			Storage:    a.Identity,
+			Clock:      a.GetClock(),
+		}); err != nil {
+			log.WithError(err).Debugf("failed U2F challenge validation using device %q", dev.Id)
```
Nit: **F**ailed

r0mant commented on lib/auth/auth.go:
```diff
@@ -843,11 +843,28 @@ func (a *Server) CheckU2FSignResponse(user string, response *u2f.AuthenticateCha
-		StorageKey: user,
-		Storage:    a.Identity,
-	})
+	ctx := context.TODO()
```
Nit: Since you're here anyway, would it make sense to refactor this method to support passing a context?

awly replied:
Propagated one layer up.
Going further will cause many more diffs and this PR is already too big

r0mant replied:
That's fair.

r0mant commented on lib/auth/init.go:
```diff
@@ -985,3 +992,67 @@ func migrateRoleOptions(ctx context.Context, asrv *Server) error {
+		}
+
+		user.SetLocalAuth(la)
+		if err := asrv.UpsertUser(user); err != nil {
```
Nit: I would consider adding a debug log entry that the user's local auth is being migrated.

alex-kovoy reviewed (approved):
It does not look that it introduces breaking changes to UI. And I assume it has been tested as well. With that, the changes look good to me!

(fyi, it might be the case that we will be adding a support for FIDO2 in the future, just to keep in mind)
